lib/spack/spack/environment.py

```diff
@@ -1509,6 +1509,45 @@ def concretized_specs(self):
         for s, h in zip(self.concretized_user_specs, self.concretized_order):
             yield (s, self.specs_by_hash[h])
 
+    def matching_spec(self, spec):
+        """
+        Given a spec (likely not concretized), find a matching concretized
+        spec in the environment. The matching spec does not have to be
+        installed in the environment.
+        """
+        root_matches = list()  # These are abstract
+        dep_matches = list()  # These are concrete
+        # These are concrete (for root specs, they store the concretized spec
+        # rather than the abstract spec)
+        all_matches = list()
+        for user_spec, concretized_user_spec in self.concretized_specs():
+            if concretized_user_spec.satisfies(spec):
+                root_matches.append(user_spec)
+                all_matches.append(concretized_user_spec)
+            for dep_spec in concretized_user_spec.traverse(root=False):
+                if dep_spec.satisfies(spec):
+                    dep_matches.append(dep_spec)
+                    all_matches.append(dep_spec)
+
+        if len(all_matches) == 1:
```

Comment:
What if there is one root match and additional dep matches? I would think that if there is only one root match, we should just return that regardless of other matches.

I also think that would allow us to make the method a little easier to read, but I'm not as sure about that.

Reply:
That makes sense so I updated it. I don't think it's simpler, but it's also not much more complex. I also updated the docstring to discuss all the semantics.